lib/pbench/server/database/models/api_keys.py

```diff
@@ -32,15 +35,38 @@ class APIKey(Database.Base):
     # Indirect reference to the owning User record
     user = relationship("User")
 
+    def __str__(self):
+        return f"key: {self.api_key}"
+
+    class DuplicateValue(commonDBException):
+        """Attempt to commit a duplicate unique value."""
+
+        def __init__(self, api_key: "APIKey", cause: str):
+            self.api_key = api_key
+            self.cause = cause
+
+        def __str__(self) -> str:
+            return f"Duplicate api_key {self.api_key}: {self.cause}"
+
+    class NullKey(commonDBException):
+        """Attempt to commit an APIkey row with an empty required column."""
+
+        def __init__(self, api_key: "APIKey", cause: str):
+            self.api_key = api_key
+            self.cause = cause
+
+        def __str__(self) -> str:
+            return f"Missing required key in {self.api_key.as_json()}: {self.cause}"
+
     def add(self):
         """Add an api_key object to the database."""
         try:
             Database.db_session.add(self)
             Database.db_session.commit()
         except Exception as e:
             Database.db_session.rollback()
-            self.logger.error("Can't add api_key to DB: {}", str(e))
-            raise APIKeyError(f"Error adding api_key to db : {e}") from e
+            self.logger.error("Can't add key:{} to DB: {}", str(self), str(e))
+            raise commonDBException._decode(self, e) from e
```

Comment:
This is problematic:  you're invoking an instance method using a class reference, and so you have to explicitly provide what is normally an implicit parameter...which should raise various red flags.  😁 

I think you should be invoking `_decode()` using an instance of (a subclass of) the `commonDBException` class; so line 69 should be something more like
```python
            if isinstance(e, commonDBException):
                raise e._decode() from e
            else:
                raise
```
or
```python
            de = e._decode() if isinstance(e, commonDBException) else e
            if de is e:
                raise
            else:
                raise de from e
```